mantidproject/mantid
# NXcanSas read and write

Description of work.

This PR implements a loader and a reader for the the NXcanSAS data format. Information regarding this data format can be found [here](http://download.nexusformat.org/sphinx/classes/contributed_definitions/NXcanSAS.html) and [here](http://cansas-org.github.io/NXcanSAS/classes/contributed_definitions/NXcanSAS.html).

Fixes #15811 
## To test:

A windows installer can be found here: \olympic\Babylon5\Scratch\Anton\Testing\NXcanSAS

The sample data can be found here: \olympic\Babylon5\Scratch\Anton\Testing\NXcanSAS\BenchMarkingSANS2D_D20

**Test saving 2D**
- Open the ISIS SANS GUI
- Set the instrument to SANS2DTUBES
- Load the user file: USER_SANS2D_154E_2p4_4m_M3_Xpress_8mm_SampleChanger.txt
- Set the input files as:

| type | Scatter | Transmission | Direct |
| --- | --- | --- | --- |
| sample | SANS2D00034484 | SANS2D00034505 | SANS2D00034461 |
| can | SANS2D00034481 | SANS2D00034502 | SANS2D00034461 |
- Press load
- Press 2D reduction

You should see an output workspace with the name: 34484rear_2D_1.75_16.5
- Open the algorithm `SaveNXcanSAS`
- Set the workspace 34484rear_2D_1.75_16.5 as the input workspace and set a file name
- Run the algorithm
  - Confirm that the file was created
  - Use HDFViewer to open the file and confirm that it can be opened
- Open the algorithm `LoadNXcanSAS`
- Set the file name to the file that was just saved
- Set the name of the output workspace to _InWS_from_NXCanSAS_
- Run the alogorithm
  - Confirm that the workspace has been loaded

Now we want to compare the workspaces. Note that 34484rear_2D_1.75_16.5 is a histogram worksapce whilie the freshly loaded workspace will be point data. 
- Apply ConvertToHistogram to 34484rear_2D_1.75_16.5 and set the output workspace name to `ReferenceWS`
- Compare the data of `ReferenceWS` and `nWS_from_NXCanSAS`
  - Confirm that the Y values are the same (compare them roughly)
  - Confirmt that the X values are the same
  - Confirm that the number of entries on what is normally the spectrum axis is the same
  - Open the SampleLog and confirm that entries for `UserFile`, `run_number` and `run_title` exist

**Test saving 1D**
1. Follow the same steps as above but press `1D Reduce` instead. You should have a 1D workspace called 34484rear_2D_1.75_16.5
2. Follow the same save and load procedure as for the 2D workspace.

**Test saving 1D with transmission workspaces**
1. Follow the same steps as above but press `1D Reduce` instead. You should have a 1D workspace called 34484rear_2D_1.75_16.5 and two transmission workspaces. Add them to the save option (one for can and one for the sample). 
2. Follow the same steps as above but enable, "Load Transmissions"



---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
##### Functional Tests
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
